From 3f6bcc094f0ddc6ecc8f855ee67db8060db4abf8 Mon Sep 17 00:00:00 2001
From: Yuwei Chen <yuwei.chen@intel.com>
Date: Wed, 29 Jul 2020 08:55:38 +0800
Subject: [Patch 03/10] Support for test case: 1.include .h and .asi in asl
 files; 2.include .asl in asl files; 3.include .h in nasm files; 4.include
 .asm in asm files; 5.include .h and .i in asm files; 6.include .inc in asm
 files;

---
 TestPkg/AcpiTables/AcpiTables.inf     |  2 ++
 TestPkg/AcpiTables/Dsdt.asl           |  8 +++----
 TestPkg/AcpiTables/Platform.h         |  1 +
 TestPkg/AcpiTables/TestPlatform.asl   | 10 +++++++++
 TestPkg/AcpiTables/test.asi           | 32 +++++++++++++++++++++++++++
 TestPkg/FakeEmuSec/Ia32/FakeTest.h    | 10 +++++++++
 TestPkg/FakeEmuSec/Ia32/FakeTest.i    | 10 +++++++++
 TestPkg/FakeEmuSec/Ia32/FakeTest.inc  | 14 ++++++++++++
 TestPkg/FakeEmuSec/Ia32/SwitchRam.asm | 18 +++++++--------
 TestPkg/FakeEmuSec/Ia32/Test.asm      | 18 +++++++++++++++
 TestPkg/FakeEmuSec/Sec.inf            |  4 ++++
 TestPkg/FakeLogo/JpegLogoDxe.inf      |  1 +
 12 files changed, 115 insertions(+), 13 deletions(-)
 create mode 100644 TestPkg/AcpiTables/TestPlatform.asl
 create mode 100644 TestPkg/AcpiTables/test.asi
 create mode 100644 TestPkg/FakeEmuSec/Ia32/FakeTest.h
 create mode 100644 TestPkg/FakeEmuSec/Ia32/FakeTest.i
 create mode 100644 TestPkg/FakeEmuSec/Ia32/FakeTest.inc
 create mode 100644 TestPkg/FakeEmuSec/Ia32/Test.asm

diff --git a/TestPkg/AcpiTables/AcpiTables.inf b/TestPkg/AcpiTables/AcpiTables.inf
index 6b3fa05fc3..ec5d9ec9f9 100644
--- a/TestPkg/AcpiTables/AcpiTables.inf
+++ b/TestPkg/AcpiTables/AcpiTables.inf
@@ -21,15 +21,17 @@
 #  VALID_ARCHITECTURES           = IA32 X64 EBC
 #
 
 [Sources]
   Platform.h
+  TestPlatform.asl
   Madt.aslc
   Facp.aslc
   Facs.aslc
   Dsdt.asl
   Ssdt.asl
+  test.asi
 
 [Packages]
   MdePkg/MdePkg.dec
   OvmfPkg/OvmfPkg.dec
 
diff --git a/TestPkg/AcpiTables/Dsdt.asl b/TestPkg/AcpiTables/Dsdt.asl
index 85f9b78d79..383d72af54 100644
--- a/TestPkg/AcpiTables/Dsdt.asl
+++ b/TestPkg/AcpiTables/Dsdt.asl
@@ -3,21 +3,21 @@
 
   Copyright (c) 2008, Intel Corporation. All rights reserved.<BR>
   SPDX-License-Identifier: BSD-2-Clause-Patent
 
 **/
-
+#include "Platform.h"
+#include "test.asi"
 DefinitionBlock ("Dsdt.aml", "DSDT", 1, "INTEL ", "OVMF    ", 4) {
   //
   // System Sleep States
   //
   // We build S3 and S4 with GetSuspendStates() in
   // "OvmfPkg/AcpiPlatformDxe/Qemu.c".
   //
   Name (\_S0, Package () {5, 0, 0, 0}) // Working
   Name (\_S5, Package () {0, 0, 0, 0}) // Soft Off
-
   //
   //  System Bus
   //
   Scope (\_SB) {
     //
@@ -80,11 +80,11 @@ DefinitionBlock ("Dsdt.aml", "DSDT", 1, "INTEL ", "OVMF    ", 4) {
           MaxFixed,              // Range is Fixed
           Cacheable,
           ReadWrite,
           0x00000000,            // Granularity
           0x000A0000,            // Min
-          0x000BFFFF,            // Max
+          MAX_DWORDMEMORY,       // Max
           0x00000000,            // Translation
           0x00020000             // Range Length
           )
 
         DWORDMEMORY (            // Descriptor for 32-bit MMIO
@@ -113,11 +113,11 @@ DefinitionBlock ("Dsdt.aml", "DSDT", 1, "INTEL ", "OVMF    ", 4) {
             MaxFixed,            // Range is Fixed
             Cacheable,
             ReadWrite,
             0x00000000,          // Granularity
             0x8000000000,        // Min
-            0xFFFFFFFFFF,        // Max
+            MAX_QWordMemory,     // Max
             0x00000000,          // Translation
             0x8000000000,        // Range Length
             ,                    // ResourceSourceIndex
             ,                    // ResourceSource
             PW64                 // DescriptorName
diff --git a/TestPkg/AcpiTables/Platform.h b/TestPkg/AcpiTables/Platform.h
index 43c2551a29..b1341ee2a4 100644
--- a/TestPkg/AcpiTables/Platform.h
+++ b/TestPkg/AcpiTables/Platform.h
@@ -50,10 +50,11 @@
                          EFI_ACPI_2_0_SLP_BUTTON | \
                          EFI_ACPI_2_0_RTC_S4 | \
                          EFI_ACPI_2_0_RESET_REG_SUP)
 #define RESET_REG       0xCF9
 #define RESET_VALUE     (BIT2 | BIT1) // PIIX3 Reset CPU + System Reset
+#define MAX_DWORDMEMORY 0x000BFFFF
 
 //
 // Byte-aligned IO port register block initializer for
 // EFI_ACPI_2_0_GENERIC_ADDRESS_STRUCTURE
 //
diff --git a/TestPkg/AcpiTables/TestPlatform.asl b/TestPkg/AcpiTables/TestPlatform.asl
new file mode 100644
index 0000000000..0f68ab8d3c
--- /dev/null
+++ b/TestPkg/AcpiTables/TestPlatform.asl
@@ -0,0 +1,10 @@
+/** @file
+  Contains root level name space objects for the platform
+
+  Copyright (c) 2008, Intel Corporation. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include "Dsdt.asl"
+
diff --git a/TestPkg/AcpiTables/test.asi b/TestPkg/AcpiTables/test.asi
new file mode 100644
index 0000000000..95554593d2
--- /dev/null
+++ b/TestPkg/AcpiTables/test.asi
@@ -0,0 +1,32 @@
+/** @file
+
+  @copyright
+  INTEL CONFIDENTIAL
+  Copyright 2018 - 2019 Intel Corporation. <BR>
+
+  The source code contained or described herein and all documents related to the
+  source code ("Material") are owned by Intel Corporation or its suppliers or
+  licensors. Title to the Material remains with Intel Corporation or its suppliers
+  and licensors. The Material may contain trade secrets and proprietary    and
+  confidential information of Intel Corporation and its suppliers and licensors,
+  and is protected by worldwide copyright and trade secret laws and treaty
+  provisions. No part of the Material may be used, copied, reproduced, modified,
+  published, uploaded, posted, transmitted, distributed, or disclosed in any way
+  without Intel's prior express written permission.
+
+  No license under any patent, copyright, trade secret or other intellectual
+  property right is granted to or conferred upon you by disclosure or delivery
+  of the Materials, either expressly, by implication, inducement, estoppel or
+  otherwise. Any license under such intellectual property rights must be
+  express and approved by Intel in writing.
+
+  Unless otherwise agreed by Intel in writing, you may not remove or alter
+  this notice or any other notice embedded in Materials by Intel or
+  Intel's suppliers or licensors in any way.
+**/
+
+
+#define MAX_QWordMemory  0xFFFFFFFFFF
+
+
+
diff --git a/TestPkg/FakeEmuSec/Ia32/FakeTest.h b/TestPkg/FakeEmuSec/Ia32/FakeTest.h
new file mode 100644
index 0000000000..7aabf4b637
--- /dev/null
+++ b/TestPkg/FakeEmuSec/Ia32/FakeTest.h
@@ -0,0 +1,10 @@
+/** @file
+
+For a test
+
+Copyright (c) 2006 - 2018, Intel Corporation. All rights reserved.<BR>
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#define testaddr1 12
\ No newline at end of file
diff --git a/TestPkg/FakeEmuSec/Ia32/FakeTest.i b/TestPkg/FakeEmuSec/Ia32/FakeTest.i
new file mode 100644
index 0000000000..ab44a529b3
--- /dev/null
+++ b/TestPkg/FakeEmuSec/Ia32/FakeTest.i
@@ -0,0 +1,10 @@
+/** @file
+
+For a test
+
+Copyright (c) 2006 - 2018, Intel Corporation. All rights reserved.<BR>
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#define testaddr2 16
\ No newline at end of file
diff --git a/TestPkg/FakeEmuSec/Ia32/FakeTest.inc b/TestPkg/FakeEmuSec/Ia32/FakeTest.inc
new file mode 100644
index 0000000000..2476b1c415
--- /dev/null
+++ b/TestPkg/FakeEmuSec/Ia32/FakeTest.inc
@@ -0,0 +1,14 @@
+;%HEADER%
+;/** @file
+;  Macros to work around lack of Apple support for LDR register, =expr
+;
+;  Copyright (c) 2009, Apple Inc. All rights reserved.<BR>
+;  Copyright (c) 2011-2012, ARM Ltd. All rights reserved.<BR>
+;
+;  SPDX-License-Identifier: BSD-2-Clause-Patent
+;
+;**/
+
+#define testaddr0 8
+ 
+END
diff --git a/TestPkg/FakeEmuSec/Ia32/SwitchRam.asm b/TestPkg/FakeEmuSec/Ia32/SwitchRam.asm
index 99ef002980..f90c569784 100644
--- a/TestPkg/FakeEmuSec/Ia32/SwitchRam.asm
+++ b/TestPkg/FakeEmuSec/Ia32/SwitchRam.asm
@@ -10,15 +10,16 @@
 ; Abstract:
 ;
 ;   Switch the stack from temporary memory to permanent memory.
 ;
 ;------------------------------------------------------------------------------
-
+#include "FakeTest.h"
+#include "FakeTest.i"
+#include "FakeTest.inc"
     .586p
     .model  flat,C
     .code
-
 ;------------------------------------------------------------------------------
 ; VOID
 ; EFIAPI
 ; SecSwitchStack (
 ;   UINT32   TemporaryMemoryBase,
@@ -31,11 +32,10 @@ SecSwitchStack   PROC
     ;
     push  eax
     push  ebx
     push  ecx
     push  edx
-
     ;
     ; !!CAUTION!! this function address's is pushed into stack after
     ; migration of whole temporary memory, so need save it to permanent
     ; memory at first!
     ;
@@ -52,16 +52,16 @@ SecSwitchStack   PROC
     add   eax, ecx
     mov   edx, dword ptr [esp]         ; copy pushed register's value to permanent memory
     mov   dword ptr [eax], edx
     mov   edx, dword ptr [esp + 4]
     mov   dword ptr [eax + 4], edx
-    mov   edx, dword ptr [esp + 8]
-    mov   dword ptr [eax + 8], edx
-    mov   edx, dword ptr [esp + 12]
-    mov   dword ptr [eax + 12], edx
-    mov   edx, dword ptr [esp + 16]    ; Update this function's return address into permanent memory
-    mov   dword ptr [eax + 16], edx
+    mov   edx, dword ptr [esp + testaddr0]
+    mov   dword ptr [eax + testaddr0], edx
+    mov   edx, dword ptr [esp + testaddr1]
+    mov   dword ptr [eax + testaddr1], edx
+    mov   edx, dword ptr [esp + testaddr2]    ; Update this function's return address into permanent memory
+    mov   dword ptr [eax + testaddr2], edx
     mov   esp, eax                     ; From now, esp is pointed to permanent memory
 
     ;
     ; Fixup the ebp point to permanent memory
     ;
diff --git a/TestPkg/FakeEmuSec/Ia32/Test.asm b/TestPkg/FakeEmuSec/Ia32/Test.asm
new file mode 100644
index 0000000000..1e692317cd
--- /dev/null
+++ b/TestPkg/FakeEmuSec/Ia32/Test.asm
@@ -0,0 +1,18 @@
+;------------------------------------------------------------------------------
+;
+; Copyright (c) 2007 - 2012, Intel Corporation. All rights reserved.<BR>
+; SPDX-License-Identifier: BSD-2-Clause-Patent
+;
+; Module Name:
+;
+;   Stack.asm
+;
+; Abstract:
+;
+;   Switch the stack from temporary memory to permanent memory.
+;
+;------------------------------------------------------------------------------
+
+#include "SwitchRam.asm"
+
+END
\ No newline at end of file
diff --git a/TestPkg/FakeEmuSec/Sec.inf b/TestPkg/FakeEmuSec/Sec.inf
index 2f9e3d4780..0a10d74c0b 100644
--- a/TestPkg/FakeEmuSec/Sec.inf
+++ b/TestPkg/FakeEmuSec/Sec.inf
@@ -26,11 +26,15 @@
   X64/SwitchRam.S
 
 [Sources.IA32]
   Ia32/TempRam.c
   Ia32/SwitchRam.asm
+  Ia32/Test.asm
   Ia32/SwitchRam.S
+  Ia32/FakeTest.h
+  Ia32/FakeTest.i
+  Ia32/FakeTest.inc
 
 [Packages]
   MdePkg/MdePkg.dec
   EmulatorPkg/EmulatorPkg.dec
 
diff --git a/TestPkg/FakeLogo/JpegLogoDxe.inf b/TestPkg/FakeLogo/JpegLogoDxe.inf
index 47f1672a4c..3c536ef0fc 100644
--- a/TestPkg/FakeLogo/JpegLogoDxe.inf
+++ b/TestPkg/FakeLogo/JpegLogoDxe.inf
@@ -38,10 +38,11 @@
 
 [LibraryClasses]
   UefiBootServicesTableLib
   UefiDriverEntryPoint
   DebugLib
+  DebugAgentLib
 
 [Protocols]
   gEfiHiiDatabaseProtocolGuid        ## CONSUMES
   gEfiHiiImageExProtocolGuid         ## CONSUMES
   gEfiHiiPackageListProtocolGuid     ## PRODUCES CONSUMES
-- 
2.29.1.windows.1

